airbyte-integrations/connectors/source-mysql-v2/src/main/kotlin/io/airbyte/integrations/source/mysql/MysqlJdbcPartitionFactory.kt

```diff
@@ -0,0 +1,145 @@
+                lowerBound = null,
+                upperBound = null,
+            )
+        }
+
+        val cursorChosenFromCatalog: Field =
+            stream.configuredCursor as? Field ?: throw ConfigErrorException("no cursor")
+
+        if (pkChosenFromCatalog.isEmpty()) {
+            return MysqlJdbcUnsplittableSnapshotWithCursorPartition(
+                selectQueryGenerator,
+                streamState,
+                cursorChosenFromCatalog
+            )
+        }
+        return MysqlJdbcSplittableSnapshotWithCursorPartition(
+            selectQueryGenerator,
+            streamState,
+            pkChosenFromCatalog,
+            lowerBound = null,
+            upperBound = null,
+            cursorChosenFromCatalog,
+            cursorUpperBound = null,
+        )
+    }
+
+    override fun create(
+        stream: Stream,
+        opaqueStateValue: OpaqueStateValue?,
+    ): MysqlJdbcPartition? {
+        val streamState: DefaultJdbcStreamState = streamState(stream)
+        if (opaqueStateValue == null) {
+            return coldStart(streamState)
+        }
+        val sv: MysqlJdbcStreamStateValue =
+            Jsons.treeToValue(opaqueStateValue, MysqlJdbcStreamStateValue::class.java)
+
+        val isCursorBasedIncremental: Boolean =
+            stream.configuredSyncMode == ConfiguredSyncMode.INCREMENTAL &&
+                !sharedState.configuration.global
+
+        if (!isCursorBasedIncremental) {
+            // todo: Implement me for cdc initial read states.
+            throw IllegalStateException("Should reset.")
+        } else {
+            if (sv.stateType != "cursor_based") {
+                // Loading value from catalog. Note there could be unexpected behaviors if user
+                // updates their schema but did not reset their state.
+                val pkChosenFromCatalog: List<Field> = stream.configuredPrimaryKey ?: listOf()
+                val pkLowerBound: JsonNode = Jsons.valueToTree(sv.pkValue)
+                val cursorChosenFromCatalog: Field =
+                    stream.configuredCursor as? Field ?: throw ConfigErrorException("no cursor")
+
+                // in a state where it's still in primary_key read part.
+                return MysqlJdbcSplittableSnapshotWithCursorPartition(
+                    selectQueryGenerator,
+                    streamState,
+                    pkChosenFromCatalog,
+                    lowerBound = listOf(pkLowerBound),
+                    upperBound = null,
+                    cursorChosenFromCatalog,
+                    cursorUpperBound = null,
+                )
+            }
+            // resume back to cursor based increment.
+            val cursor: Field = stream.fields.find { it.id == sv.cursorField.first() } as Field
+            // Compose a jsonnode of cursor label to cursor value to fit in
+            // DefaultJdbcCursorIncrementalPartition
+            val cursorCheckpoint: JsonNode = Jsons.valueToTree(sv.cursors)
```

Comment:
@postamar A question to you: in v1 all cursor values are stored as json string (like `"cursor": "2"` where the cursor is actually integer 2).

I think we need to cast this back to integer before converting this back to JsonNode, otherwise IntCodec would accept it, but with current implementation I don't find a good way to do so. I feel we are missing a map between `Airbytetype` (or `FieldType`) to actual kotlin type. What do you think?

Reply:
Interesting. Would `Jsons.readTree(sv.cursors)` do the trick? This will parse a `JsonNode` from a string. Here `Jsons.valueToTree(sv.cursors)` will simply wrap the string in a `TextNode` which, regardless of what the solution to the problem actually is, is probably not what we want to do.

Reply:
I think it would be related to the type of the cursor. In v1 state, all cursor value regardless of the type would be converted to a string. Thus, reading it back from a previous state would cause trouble for Intcodec because it would expect `2` rather than `"2"`.

Reply:
fixed. thanks for walking me through it!